### DISCUSSION TITLE
Is C# FomattableString supported in PWSH language?

### DISCUSSION THREAD

**Discussion by netcorefan1** [PowerShell/PowerShell#19670] [Q&A] [answered]
Hello,
I'm trying to call [this](https://spectreconsole.net/api/spectre.console/ansiconsole/578156ab) C# function which requires a [FormattableString](https://learn.microsoft.com/en-us/dotnet/api/system.formattablestring) as parameter, but without success. I'm wondering if I'm doing something wrong or if this is just a limitation of the PWSH language itself.
```
$culture = [System.Globalization.CultureInfo]::InvariantCulture
$input  = "my input"
[System.FormattableString] $fs = "${input}"
[System.FormattableString] $fs = "$input"
[System.FormattableString] $fs = "$($input)"
[AnsiConsole]::MarkupLineInterpolated($culture, $fs)
```
`MetadataError: Cannot convert the "my input" value of type "System.String" to type "System.FormattableString".`

**Comment by rhubarb-geek-nz** [comment, marked as answer]
The C#/dotnet layer has no knowledge of PowerShell variables.

```
[String]::Format("{0},{2}",0,1,2)
[System.Console]::WriteLine("{2},{1}",0,1,2)
```

gives

```
0,2
2,1
```

So

```
$input = "my input"
[System.Console]::WriteLine("{0}",$input)
```

gives

```
my input
```

Also, [System.FormattableString](https://learn.microsoft.com/en-us/dotnet/api/system.formattablestring.-ctor?view=net-7.0#system-formattablestring-ctor) has no constructor that takes a value. It is actually constructed at compile time.

Try [FormattableStringFactory.Create](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.formattablestringfactory.create?view=net-6.0#system-runtime-compilerservices-formattablestringfactory-create(system-string-system-object()))

**Reply by netcorefan1** [reply]
> Also, [System.FormattableString](https://learn.microsoft.com/en-us/dotnet/api/system.formattablestring.-ctor?view=net-7.0#system-formattablestring-ctor) has no constructor that takes a value. It is actually constructed at `compile time.`
> 
> Try [FormattableStringFactory.Create](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.formattablestringfactory.create?view=net-6.0#system-runtime-compilerservices-formattablestringfactory-create(system-string-system-object()))

Many thanks. @rhubarb-geek-nz ! It worked:
```
$input  = "999"
[AnsiConsole]::MarkupLineInterpolated([FormattableStringFactory]::Create("My input value is: {0}", $input))
```
Is there a way to do a string interpolation at compile time in PWSH?
I suppose that creating a `FormattableString` at runtime each time there is something to print on the screen is not good for performance, but at least now I can call `MarkupLineInterpolated` rather than having to reimplement the entire function in PS.

**Reply by SeeminglyScience** [reply]
> Is there a way to do a string interpolation at compile time in PWSH?

A lot of the reasons that `FormattableStringFactory` was created don't apply quite as heavily to PowerShell. Primarily because it's all dynamic so most things get boxed at some point anyway. I'd just stick with a normal `[AnsiConsole]::MarkupLine("My input value is: $something")`. It'll do `string.Format` behind the scenes which isn't quite as fancy but it'll be fine in the vast majority of scenarios.
